### PR TITLE
feat(derive): BatchQueue Implementation

### LINKED ISSUE #52: feat(derive): Batch Queue stage
## Overview

The `Batch Queue` is responsible for reordering batches by their timestamps. It also fills in gaps by creating empty batches if any are missing for certain time slots.

**References**
* [Batch Queue](https://specs.optimism.io/protocol/derivation.html#batch-queue) specification.

### PR DESCRIPTION
**Pull request by refcell**
**Description**

Implements the `BatchQueue`.

**Metadata**

Fixes #52

### CODE REVIEW

File: crates/derive/src/traits/data_sources.rs [hunk 1]
```diff
@@ -20,6 +20,18 @@ pub trait ChainProvider {
     async fn receipts_by_hash(&self, hash: B256) -> Result<Vec<Receipt>>;
 }
 
+/// Describes the functionality of a data source that fetches safe blocks.
+#[async_trait]
+pub trait SafeBlockFetcher {
+    /// Returns the block reference (info) given a block number.
+    /// Should error if the block does not exist.
+    async fn block_info_by_number(&self, number: u64) -> Result<BlockInfo>;
```
Comment (clabby) [inline crates/derive/src/traits/data_sources.rs:28]: Should be `L2BlockInfo`. Maybe rename to `l2_block_info_by_number` as well? `block_info_by_number` is already in `ChainProvider`

File: crates/derive/src/stages/batch_queue.rs [hunk 2]
```diff
@@ -1 +1,369 @@
...
+    pub fn origin(&self) -> Option<&BlockInfo> {
+        self.prev.origin()
+    }
+
+    /// Pops the next batch from the current queued up span-batch cache.
+    /// The parent is used to set the parent hash of the batch.
+    /// The parent is verified when the batch is later validated.
+    pub fn pop_next_batch(&mut self, parent: L2BlockRef) -> Option<SingleBatch> {
+        if self.next_spans.is_empty() {
+            return None;
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:91]: Should panic here. Invalid state, derivation cannot continue.

File: crates/derive/src/stages/batch_queue.rs [hunk 3]
```diff
@@ -1 +1,369 @@
...
+
+        // If the epoch is advanced, update the l1 blocks.
+        // Advancing epoch must be done after the pipeline successfully applies the entire span
+        // batch to the chain.
+        // Because the span batch can be reverted during processing the batch, then we must
+        // preserve existing l1 blocks to verify the epochs of the next candidate batch.
+        if !self.l1_blocks.is_empty() && parent.l1_origin.number > self.l1_blocks[0].number {
+            for (i, block) in self.l1_blocks.iter().enumerate() {
+                if parent.l1_origin.number == block.number {
+                    self.l1_blocks.drain(0..=i);
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:125]: Think we'd like to keep `i` here
```suggestion
                    self.l1_blocks.drain(0..i);
```

File: crates/derive/src/stages/batch_queue.rs [hunk 4]
```diff
@@ -1 +1,369 @@
...
+        let origin_behind =
+            self.origin.map_or(true, |origin| origin.number < parent.l1_origin.number);
+
+        // Advance the origin if needed.
+        // The entire pipeline has the same origin.
+        // Batches prior to the l1 origin of the l2 safe head are not accepted.
+        if self.origin != self.prev.origin().copied() {
+            self.origin = self.prev.origin().cloned();
+            if !origin_behind {
+                self.l1_blocks.push(*self.origin.as_ref().unwrap());
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:147]: Handle `unwrap` here?

File: crates/derive/src/stages/batch_queue.rs [hunk 5]
```diff
@@ -1 +1,369 @@
...
+            Err(e) => return Err(e),
+        }
+
+        // Skip adding the data unless up to date with the origin,
+        // but still fully empty the previous stages.
+        if origin_behind {
+            if out_of_data {
+                return Err(StageError::Eof);
+            }
+            return Err(StageError::Custom(anyhow!("Not Enough Data")));
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:178]: Would be nice to add a `NotEnoughData` variant, used several times

File: crates/derive/src/stages/batch_queue.rs [hunk 6]
```diff
@@ -1 +1,369 @@
...
+                _ => return Err(e),
+            },
+        };
+
+        // If the next batch is derived from the span batch, it's the last batch of the span.
+        // For singular batches, the span batch cache should be empty.
+        match batch {
+            Batch::Single(sb) => Ok(sb),
+            Batch::Span(sb) => {
+                let batches = sb.get_singular_batches(&self.l1_blocks, parent); // .ok_or_else(|| anyhow!("failed to get singular batches from span batch"))?;
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:200]: Delete comment?

File: crates/derive/src/stages/batch_queue.rs [hunk 7]
```diff
@@ -1 +1,369 @@
...
+        // bq.log.Trace("Advancing internal L1 blocks", "next_timestamp", nextTimestamp,
+        // "next_epoch_time", nextEpoch.Time)
+        self.l1_blocks.remove(0);
+        Err(StageError::Eof)
+    }
+
+    /// Adds a batch to the queue.
+    pub fn add_batch(&mut self, batch: Batch, parent: L2BlockRef) -> StageResult<()> {
+        if self.l1_blocks.is_empty() {
+            return Err(anyhow!(
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:331]: Go impl panics here, unreachable case

File: crates/derive/src/types/block.rs [hunk 8]
```diff
@@ -24,6 +24,52 @@ impl BlockInfo {
...
+impl core::fmt::Display for BlockID {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{{ hash: {}, number: {} }}", self.hash, self.number)
+    }
+}
+
+/// An L2 Block Ref
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct L2BlockRef {
```
Comment (clabby) [inline crates/derive/src/types/block.rs:63]: Looks like some types got dup'd when doing parallel work. Can we replace the `L2BlockInfo` type w/ this or vice versa?

File: crates/derive/src/stages/batch_queue.rs (hunk 3 above)
Comment (refcell) [reply crates/derive/src/stages/batch_queue.rs:125]: Good catch!

File: crates/derive/src/types/block.rs (hunk 8 above)
Comment (refcell) [reply crates/derive/src/types/block.rs:63]: Ah missed that, thanks! Bad rebase/merge

**Review by clabby — APPROVED** [review]

**Pull request merged**